rest-api/dao/participant_dao.py

```diff
@@ -114,25 +114,43 @@ def _do_update(self, session, obj, existing_obj):
                             else None)
       need_new_summary = True
 
-    if obj.organizationId or obj.siteId or obj.hpoId:
-      site, organization, awardee = self.get_pairing_level(obj)
-      obj.organizationId = organization
-      obj.siteId = site
-      obj.hpoId = awardee
-      if awardee:
-        # get provider link for hpo_id (awardee)
-        obj.providerLink = make_primary_provider_link_for_id(awardee)
+    update_pairing = True
+    if obj.siteId is None and obj.organizationId is None and obj.hpoId is None and \
+      obj.providerLink == 'null':
+      update_pairing = False
+
+    if update_pairing == True:
+      if obj.organizationId or obj.siteId or obj.hpoId:
+        site, organization, awardee = self.get_pairing_level(obj)
+        obj.organizationId = organization
+        obj.siteId = site
+        obj.hpoId = awardee
+        if awardee is not None and (obj.hpoId != existing_obj.hpoId):
+          # get provider link for hpo_id (awardee)
+          obj.providerLink = make_primary_provider_link_for_id(awardee)
 
-      need_new_summary = True
+        need_new_summary = True
 
-    # If the provider link changes, update the HPO ID on the participant and its summary.
-    if obj.hpoId is None:
+    if update_pairing == True:
+      # If the provider link changes, update the HPO ID on the participant and its summary.
+      if obj.hpoId is None:
+        obj.hpoId = existing_obj.hpoId
+      if obj.providerLink != existing_obj.providerLink:
+        new_hpo_id = self._get_hpo_id(obj)
+        if new_hpo_id != existing_obj.hpoId:
+          obj.hpoId = new_hpo_id
+          if obj.hpoId == 0:
```

Comment:
1) We should use UNSET, not 0.

2) I don't think this logic is quite right. If you started out with pairing at the org or site level, and then passed up a providerLink for another awardee, site and organization should be cleared out... but they aren't. This could lead you to having a providerLink and hpoId that doesn't match the organizationId and siteId. (Please add a test for this.) You could remove this if, but if you do that you'll wind up clearing these out when you set providerLink on line 130.

I think the right thing to do is check for providerLink being set / changing before you get to line 122, and if it is, skip lines 122 -> 130.

You could alternatively change get_pairing_level so that it looks at providerLink first, before looking at hpoId / organizationId / siteId. This would mean that if you pass up an old providerLink with other values for hpoId / organizationId / siteId, the latter would be ignored -- but I think that's probably OK. I have told them to stop passing up providerLink when they do pairing with the new system.